relayer/src/chain/cosmos/version.rs

```diff
@@ -142,6 +155,37 @@ fn parse_ibc_go_version(version_info: &VersionInfo) -> Result<Option<semver::Ver
     }
 }
 
+fn parse_tendermint_version(version_info: &VersionInfo) -> Result<semver::Version, Error> {
+    let module = version_info
+        .build_deps
+        .iter()
+        .find(|&m| m.path.contains(TENDERMINT_MODULE_NAME))
+        .ok_or_else(|| {
+            Error::tendermint_module_not_found(
+                TENDERMINT_MODULE_NAME.to_string(),
+                AppInfo::from(version_info),
+            )
+        })?;
+
+    // The raw version number has a leading 'v', trim it out;
+    let plain_version = module.version.trim_start_matches('v');
+
+    // Parse the module version
+    let mut version = semver::Version::parse(plain_version).map_err(|e| {
+        Error::version_parsing_failed(
+            module.path.clone(),
+            module.version.clone(),
+            e.to_string(),
+            AppInfo::from(version_info),
+        )
+    })?;
+
+    // Remove the pre-release version to ensure we don't give special treatment to pre-releases.
+    version.pre = semver::Prerelease::EMPTY;
```

Comment:
This could be problematic, since the -pre versions semantically precede the release without the pre suffix. So this is lossily bumping the version to a later one than what the module author intended.

Reply:
> The issue was that networks were using an RC instead of the stable release, e.g.
> 
> SDK module at version '0.42.0-rc0' does not meet compatibility requirements >=0.41.3, <=0.42.6

This semver check behavior is stricter than what I assumed. I expected that 0.42.0-rc0 does succeed 0.41.3 and precede 0.42.6, but would not meet this requirement: >= 0.42.0. Are pre-release suffixes treated as completely incomparable by the comparison logic?

> Which was fixed by trimming the `Prerelease` element. Not sure if you're hinting at a different issue than the one we noticed in the production signaled in #1337.

Thanks for the exposition on this. A proper solution, I think, would be for the compatibility requirements to explicitly include any pre-releases they want to allow, but perhaps practicalities dictate otherwise.

Reply:
> Do pre-release suffixes make versions completely incomparable by the comparison logic?

We could give this a try in a small experiment. Not sure what you want to try exactly, but I managed to reproduce the original issue here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c31232aaa272967f2979abf71e83f258

> A proper solution, I think, would be for the compatibility requirements to explicitly include any pre-releases they want to allow, but perhaps practicalities dictate otherwise.

Indeed, it's not feasible for us to track all pre-releases and test with each of them individually. We assume that pre-releases have the same API as their corresponding release.

Reply:
Thanks for the playground; it appears that the `semver` crate does treat non-empty pre-release segment as not matching any comparisons with regular versions:

> thread 'main' panicked at 'version 0.42.0-rc0 matching req >0.41.5', src/main.rs:14:5

The logic of `VersionReq::match` is poorly documented in [semver docs](https://docs.rs/semver/1.0.10/semver/index.html), even though the `Version` type as such implements a total ordering that _is_ documented to work in the way I described.
Then there is this:

> Where the various tools differ in their interpretation or implementation of the spec, this crate follows the implementation choices made by Cargo. If you are operating on version numbers from some other package ecosystem, you will want to use a different semver library which is appropriate to that ecosystem.